### ISSUE TITLE
Transformers are RNNs: Fast Autoregressive Transformers with Linear Attention

### ISSUE THREAD

**Issue by reo11** [reo11/papers100knock#55]
## 概要
- Transformerの計算量をO(N^2)からO(N)にした
- recurrent neural networksとの関係を示す
- 元のTransformerと変わらぬ性能で4,000倍高速化

## 投稿年/会議
2020年/ICML2020
- https://arxiv.org/abs/2006.16236

## 著者/所属機関
Angelos Katharopoulos, Apoorv Vyas, Nikolaos Pappas, François Fleuret
- Idiap Research Institute
- EPFL
- University of Washington

## 参考リンク
- https://www.youtube.com/watch?v=hAooAOFRsYc

### Linearized Attention
- normal transformer
![image](https://user-images.githubusercontent.com/38539915/86535515-d8c47f80-bf1b-11ea-8bb0-3bafdfe1f59c.png)

- QとKの評価関数をsimとする
![image](https://user-images.githubusercontent.com/38539915/86536365-dbc26e80-bf21-11ea-8ca0-a76f7a6ad8a8.png)
- 通常のtransformerの場合以下
![image](https://user-images.githubusercontent.com/38539915/86536379-fe548780-bf21-11ea-9994-224d99a09695.png)

- Linearized Attention
- 線形な関数φでsimを置き換えるとQとKについて分離可能
  - このφはfeature representationと書いてあって高次元空間への写像？
![image](https://user-images.githubusercontent.com/38539915/86536369-e0872280-bf21-11ea-822b-bc9ffbb8460c.png)

- Q_iをΣの外に出すことでjについて前計算しておけば計算量が大幅に減少（O(N)）
![image](https://user-images.githubusercontent.com/38539915/86535534-f72a7b00-bf1b-11ea-9ff0-2a419189ef90.png)

### Causal Masking
- RNNのようにauto regressiveに学習をする場合
![image](https://user-images.githubusercontent.com/38539915/86536550-85eec600-bf23-11ea-9bb2-d343c56ce00d.png)

![image](https://user-images.githubusercontent.com/38539915/86536559-9737d280-bf23-11ea-87d8-5babc184ec01.png)

![image](https://user-images.githubusercontent.com/38539915/86536561-9acb5980-bf23-11ea-96ff-d614cf51e2d7.png)

![image](https://user-images.githubusercontent.com/38539915/86536614-f269c500-bf23-11ea-967b-34e490467f6b.png)

- S_iとZ_iはそれまでのS_{i-1}とZ_{i-1}に対してi番目のkeyを計算するだけで良いので線形時間で学習可能

### Transformers are RNNs
- i-1までの全てを計算しなおす必要がないので、TransformerをRNNのように使える
![image](https://user-images.githubusercontent.com/38539915/86536669-5ab8a680-bf24-11ea-8388-3272511c0cb4.png)

### Results
- 赤：Transformer
- 青：Reformer (lsh-1, 4, 8)
- 黒：Linear Transformer
![image](https://user-images.githubusercontent.com/38539915/86536695-95bada00-bf24-11ea-8da1-2dd592e53333.png)
- 時間、空間計算量共にTransformer, Reformerより改善
- O(NlogN)でもO(N)でも目に見える変化はない（logは定数定期）

- Lossの減り方
  - Reformerより最終的なlossはオリジナルに近くなっている
![image](https://user-images.githubusercontent.com/38539915/86536767-1548a900-bf25-11ea-9275-d85f24988ff6.png)

#### Autoregresive image generation
- MNIST
![image](https://user-images.githubusercontent.com/38539915/86536783-44f7b100-bf25-11ea-8373-514d49d933db.png)
![image](https://user-images.githubusercontent.com/38539915/86536807-740e2280-bf25-11ea-92a5-0081904654ab.png)

- CIFAR-10
![image](https://user-images.githubusercontent.com/38539915/86536785-4de88280-bf25-11ea-9b59-3364001e20a2.png)
![image](https://user-images.githubusercontent.com/38539915/86536808-78d2d680-bf25-11ea-88b9-45129077850a.png)

### Automatic speech recognition
- WSJ datset
![image](https://user-images.githubusercontent.com/38539915/86536818-99029580-bf25-11ea-8636-0a3f16b45ea7.png)
- Bi-LSTM, Reformer (LSH-4)は上回る
- Original (Softmax)にはおよばず
  タスクによっては精度が低いが、計算効率は大幅に改善されている
